lib/monitor.toit

```diff
@@ -53,81 +56,109 @@ monitor Latch:
+*/
+monitor Semaphore:
+  count_ /int := 0
+  limit_ /int?
 
-  initializer_/Lambda? := ?
-  state_ := STATE_UNINITIALIZED_
-  value_ := null
+  /**
+  Constructs a semaphore with an initial $count and an optional $limit.
+  */
+  constructor --count/int=0 --limit/int?=null:
+    if limit and (limit < 1 or count > limit): throw "INVALID_ARGUMENT"
+    limit_ = limit
 
   /**
-  Constructs the once monitor with the given $initializer_.
+  Increments an internal counter.
 
-  The $initializer_ must be a lambda.
+  Originally called the V operation.
   */
-  constructor .initializer_:
-    if not initializer_: throw "invalid argument"
+  up -> none:
+    count := count_
+    limit := limit_
+    if limit and count >= limit: return
+    count_ = count + 1
 
   /**
-  Gets the result of the initialization.
+  Decrements an internal counter.
+  This method blocks until the counter is non-zero.
 
-  Calls the initializer the first time this method is called.
-  If the initialization throws an exception, then this methods throws that
-    exception. Future attempts to get the value will run the initializer again.
-  The trace of the initialization is printed but not rethrown.
+  Originally called the P operation.
   */
-  get:
-    if initializer_:
-      exception := catch --trace: value_ = initializer_.call
-      initializer_ = null
-      if exception:
-        state_ = STATE_EXCEPTION_
-        value_ = exception
-      else:
-        state_ = STATE_INITIALIZED_
-    if state_ == STATE_INITIALIZED_: return value_
-    else: throw value_
+  down -> none:
+    await: count_ > 0
+    count_--
 
 /**
-A semaphore synchronization primitive.
+A signal synchronization primitive.
 
 # Inheritance
 This class must not be extended.
 */
-monitor Semaphore:
+monitor Signal:
+  current_ /int := 0
+  awaited_ /int := 0
+
   /**
-  Increments an internal counter.
+  Waits until the signal has been raised while this task has been waiting.
 
-  Originally called the V operation.
+  Raises that occur before $wait has been called are not taken into
+    account, so care must be taken to avoid losing information.
   */
-  up:
-    count_++
+  wait -> none:
+    wait_: true
 
   /**
-  Decrements an internal counter.
-  This method blocks until the counter is non-zero.
+  Waits until the given $condition returns true.
 
-  Originally called the P operation.
+  The $condition is evaluated on entry, but if it returns false initially, it
+    is only re-evaluated when the signal has been raised and the current task
+    has been woken up.
   */
-  down:
-    await: count_ > 0
-    count_--
+  wait [condition] -> none:
+    if condition.call: return
+    wait_ condition
+
+  /**
+  Raises the signal and unblocks the tasks that are already waiting.
 
-  count_ := 0
+  If $max is provided and not null, no more than $max tasks are
+    woken up in the order in which they started waiting (FIFO).
+    The most common use case is to wake waiters up one at a time.
```

Comment:
The current implementation only runs the condition of `max` waiters.

It feels like actually waking `max` waiters makes more sense. I do realize that complicates the `wait_` method, though.

Reply:
Not sure. I think the max parameter is useful to control the amount of re-evaluation -- not unblock a particular number of tasks. I could probably be convinced otherwise :)